#### Kutuphanelerin Yuklenmesi

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('WA_Fn-UseC_-Telco-Customer-Churn.csv')

#### Veri Setini Anlama

In [3]:
df.head()

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity  ... DeviceProtection  \
0  No phone service             DSL             No  ...               No   
1                No             DSL            Yes  ...              Yes   
2                No             DSL            Yes  ...               No   
3  No phone service             DSL            Yes  ...              Yes   
4                No     Fiber optic             No  ...               No   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0          No          No              No  Month-to-month              Yes   
1          No          No              No        One year               No   
2          No          No              No  Month-to-month              Yes   
3         Yes          No              No        One year               No   
4          No          No              No  Month-to-month              Yes   

               PaymentMethod MonthlyCharges  TotalCharges Churn  
0           Electronic check          29.85         29.85    No  
1               Mailed check          56.95        1889.5    No  
2               Mailed check          53.85        108.15   Yes  
3  Bank transfer (automatic)          42.30       1840.75    No  
4           Electronic check          70.70        151.65   Yes  

[5 rows x 21 columns]

5 Satir 21 Kolon var. Kolon satir donusturmesi (Transpose)

In [4]:
df.T

0             1               2     \
customerID              7590-VHVEG    5575-GNVDE      3668-QPYBK   
gender                      Female          Male            Male   
SeniorCitizen                    0             0               0   
Partner                        Yes            No              No   
Dependents                      No            No              No   
tenure                           1            34               2   
PhoneService                    No           Yes             Yes   
MultipleLines     No phone service            No              No   
InternetService                DSL           DSL             DSL   
OnlineSecurity                  No           Yes             Yes   
OnlineBackup                   Yes            No             Yes   
DeviceProtection                No           Yes              No   
TechSupport                     No            No              No   
StreamingTV                     No            No              No   
StreamingMovies                 No            No              No   
Contract            Month-to-month      One year  Month-to-month   
PaperlessBilling               Yes            No             Yes   
PaymentMethod     Electronic check  Mailed check    Mailed check   
MonthlyCharges               29.85         56.95           53.85   
TotalCharges                 29.85        1889.5          108.15   
Churn                           No            No             Yes   

                                       3                 4     \
customerID                       7795-CFOCW        9237-HQITU   
gender                                 Male            Female   
SeniorCitizen                             0                 0   
Partner                                  No                No   
Dependents                               No                No   
tenure                                   45                 2   
PhoneService                             No               Yes   
MultipleLines              No phone service                No   
InternetService                         DSL       Fiber optic   
OnlineSecurity                          Yes                No   
OnlineBackup                             No                No   
DeviceProtection                        Yes                No   
TechSupport                             Yes                No   
StreamingTV                              No                No   
StreamingMovies                          No                No   
Contract                           One year    Month-to-month   
PaperlessBilling                         No               Yes   
PaymentMethod     Bank transfer (automatic)  Electronic check   
MonthlyCharges                         42.3              70.7   
TotalCharges                        1840.75            151.65   
Churn                                    No               Yes   

                              5                        6                 7     \
customerID              9305-CDSKC               1452-KIOVK        6713-OKOMC   
gender                      Female                     Male            Female   
SeniorCitizen                    0                        0                 0   
Partner                         No                       No                No   
Dependents                      No                      Yes                No   
tenure                           8                       22                10   
PhoneService                   Yes                      Yes                No   
MultipleLines                  Yes                      Yes  No phone service   
InternetService        Fiber optic              Fiber optic               DSL   
OnlineSecurity                  No                       No               Yes   
OnlineBackup                    No                      Yes                No   
DeviceProtection               Yes                       No                No   
TechSupport                     No               

Hedef (Target) Degisken "Churn" Sutunu, Sonuc (Bagimli) Degisken, Diger Degiskenler oznitelik olarak ifade edilir.

In [5]:
df.shape

(7043, 21)

7043 Orneklem (Satir)
21   Degisken (Satir)

#### Veri Onisleme

In [6]:
df.dtypes

customerID           object
gender               object
SeniorCitizen         int64
Partner              object
Dependents           object
tenure                int64
PhoneService         object
MultipleLines        object
InternetService      object
OnlineSecurity       object
OnlineBackup         object
DeviceProtection     object
TechSupport          object
StreamingTV          object
StreamingMovies      object
Contract             object
PaperlessBilling     object
PaymentMethod        object
MonthlyCharges      float64
TotalCharges         object
Churn                object
dtype: object

- 2 integer, 1 float ve 18 string degisken
- SeniorCitizen sayisal olmasina ragmen 1 ve 0 dan olustugu icin aslinda kategoriktir.
- TotalCharges string, int olmali

In [7]:
df.TotalCharges = pd.to_numeric(df.TotalCharges, errors='coerce')

- errors='coerse' ----> sayisal olmayan degerlere NaN degere atama

Eksik veri tespiti

In [8]:
df.isnull().sum()

customerID           0
gender               0
SeniorCitizen        0
Partner              0
Dependents           0
tenure               0
PhoneService         0
MultipleLines        0
InternetService      0
OnlineSecurity       0
OnlineBackup         0
DeviceProtection     0
TechSupport          0
StreamingTV          0
StreamingMovies      0
Contract             0
PaperlessBilling     0
PaymentMethod        0
MonthlyCharges       0
TotalCharges        11
Churn                0
dtype: int64

In [9]:
mean = df['TotalCharges'].mean()

In [10]:
df.TotalCharges = df.TotalCharges.fillna(mean)

In [11]:
df.isnull().sum()

customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64

In [12]:
df.isnull().sum().sum()

0

Sutun Adlarinda duzenleme
- butun karakterler kucuk harf olsun
- bosluk olan yerlere '_' yerlestirelim

In [13]:
df.columns = df.columns.str.lower().str.replace(' ', '_')

Tablodaki degerler uzerinde string optimizasyonu ve standartizasyonu

In [14]:
string_columns = list(df.dtypes[df.dtypes == 'object'].index)

In [15]:
string_columns

['customerid',
 'gender',
 'partner',
 'dependents',
 'phoneservice',
 'multiplelines',
 'internetservice',
 'onlinesecurity',
 'onlinebackup',
 'deviceprotection',
 'techsupport',
 'streamingtv',
 'streamingmovies',
 'contract',
 'paperlessbilling',
 'paymentmethod',
 'churn']

In [16]:
for col in string_columns:
    df[col]=df[col].str.lower().str.replace(' ', '_')

In [17]:
df.head(10)

customerid  gender  seniorcitizen partner dependents  tenure phoneservice  \
0  7590-vhveg  female              0     yes         no       1           no   
1  5575-gnvde    male              0      no         no      34          yes   
2  3668-qpybk    male              0      no         no       2          yes   
3  7795-cfocw    male              0      no         no      45           no   
4  9237-hqitu  female              0      no         no       2          yes   
5  9305-cdskc  female              0      no         no       8          yes   
6  1452-kiovk    male              0      no        yes      22          yes   
7  6713-okomc  female              0      no         no      10           no   
8  7892-pookp  female              0     yes         no      28          yes   
9  6388-tabgu    male              0      no        yes      62          yes   

      multiplelines internetservice onlinesecurity  ... deviceprotection  \
0  no_phone_service             dsl             no  ...               no   
1                no             dsl            yes  ...              yes   
2                no             dsl            yes  ...               no   
3  no_phone_service             dsl            yes  ...              yes   
4                no     fiber_optic             no  ...               no   
5               yes     fiber_optic             no  ...              yes   
6               yes     fiber_optic             no  ...               no   
7  no_phone_service             dsl            yes  ...               no   
8               yes     fiber_optic             no  ...              yes   
9                no             dsl            yes  ...               no   

  techsupport streamingtv streamingmovies        contract paperlessbilling  \
0          no          no              no  month-to-month              yes   
1          no          no              no        one_year               no   
2          no          no              no  month-to-month              yes   
3         yes          no              no        one_year               no   
4          no          no              no  month-to-month              yes   
5          no         yes             yes  month-to-month              yes   
6          no         yes              no  month-to-month              yes   
7          no          no              no  month-to-month               no   
8         yes         yes             yes  month-to-month              yes   
9          no          no              no        one_year               no   

               paymentmethod monthlycharges  totalcharges  churn  
0           electronic_check          29.85         29.85     no  
1               mailed_check          56.95       1889.50     no  
2               mailed_check          53.85        108.15    yes  
3  bank_transfer_(automatic)          42.30       1840.75     no  
4           electronic_check          70.70        151.65    yes  
5           electronic_check          99.65        820.50    yes  
6    credit_card_(automatic)          89.10       1949.40     no  
7               mailed_check          29.75        301.90     no  
8           electronic_check         104.80       3046.05    yes  
9  bank_transfer_(automatic)          56.15       3487.95     no  

[10 rows x 21 columns]

Hedef degiskende 'churn' duzenleme 

In [18]:
df.churn.head()

0     no
1     no
2    yes
3     no
4    yes
Name: churn, dtype: object

- yes ----> musteri servisi birakmis (abonelikten cikmis)
- no  ----> musteri servisi birakmamis (abonelikten cikmamis)

makine ogrenmesinde (ozellikle sklearn kutuphanesi) sayisal degerler kullanilir.

In [19]:
df.churn = (df.churn == 'yes').astype(int)

In [20]:
df.churn.head()

0    0
1    0
2    1
3    0
4    1
Name: churn, dtype: int64

In [21]:
df.head()

customerid  gender  seniorcitizen partner dependents  tenure phoneservice  \
0  7590-vhveg  female              0     yes         no       1           no   
1  5575-gnvde    male              0      no         no      34          yes   
2  3668-qpybk    male              0      no         no       2          yes   
3  7795-cfocw    male              0      no         no      45           no   
4  9237-hqitu  female              0      no         no       2          yes   

      multiplelines internetservice onlinesecurity  ... deviceprotection  \
0  no_phone_service             dsl             no  ...               no   
1                no             dsl            yes  ...              yes   
2                no             dsl            yes  ...               no   
3  no_phone_service             dsl            yes  ...              yes   
4                no     fiber_optic             no  ...               no   

  techsupport streamingtv streamingmovies        contract paperlessbilling  \
0          no          no              no  month-to-month              yes   
1          no          no              no        one_year               no   
2          no          no              no  month-to-month              yes   
3         yes          no              no        one_year               no   
4          no          no              no  month-to-month              yes   

               paymentmethod monthlycharges  totalcharges  churn  
0           electronic_check          29.85         29.85      0  
1               mailed_check          56.95       1889.50      0  
2               mailed_check          53.85        108.15      1  
3  bank_transfer_(automatic)          42.30       1840.75      0  
4           electronic_check          70.70        151.65      1  

[5 rows x 21 columns]

- Veri Setinin Egitim(Train) ve Test olarak Bolunmesi
- Train ile model kurulur. Test ille model degerlendirilir.

In [22]:
from sklearn.model_selection import train_test_split

In [23]:
df_train_full, df_test = train_test_split(df, test_size=0.2, random_state=40)

"random_state" İşlevdeki parametre, "train_test_split" testi bölünmesinin yeniden üretilebilir olmasını sağlayan rasgele sayı üreteci için rasgele çekirdeği ayarlamak için kullanılır.

parametre random_stateolarak ayarlanır 42. Bu, bu kodu her çalıştırdığınızda, rasgele sayı üreteci her zaman aynı çekirdeği kullanacağından, tren testi bölümünün aynı olacağı anlamına gelir.

random_stateÖrneğin, farklı modelleri veya parametreleri test ederken sonuçların tekrar üretilebilir olduğundan emin olmak istediğinizde, öğesini sabit bir değere ayarlamak yararlı olabilir.

Dolayısıyla, değer 42olarak seçim yapmak random_statedoğru ya da yanlış değil, veri bilimi topluluğunda popüler hale gelen bir kuraldır. Diğer herhangi bir değerden doğası gereği daha iyi veya daha kötü değildir, ancak sabit bir değer kullanmak sonuçların tekrarlanabilirliğini sağlamaya yardımcı olabilir.

#### validation verisi olusturma

train verisinin bir kismini validation verisi olarak ayarlama

In [24]:
df_train, df_val = train_test_split(df_train_full, test_size=0.25, random_state=40)

#### Hedef Degisken ve Oznitelikler

Hedef Sutun ----> churn

In [25]:
y_train = df_train.churn.values

validation icin hedef degisken

In [26]:
y_val = df_val.churn.values

hedef sutununun veri setinden silinmesi

In [27]:
del df_train['churn']
del df_val['churn']

Oznitelik Muhendisligi (Feature Engineering)

In [33]:
categorical = [
 'gender',
 'seniorcitizen',
 'partner',
 'dependents',
 'phoneservice',
 'multiplelines',
 'internetservice',
 'onlinesecurity',
 'onlinebackup',
 'deviceprotection',
 'techsupport',
 'streamingtv',
 'streamingmovies',
 'contract',
 'paperlessbilling',
 'paymentmethod',
 ]

In [34]:
numerical = ['tenure', 'monthlycharges', 'totalcharges']

In [36]:
train_dict = df_train[categorical + numerical].to_dict(orient='records')

In [37]:
train_dict[0]

{'gender': 'female',
 'seniorcitizen': 0,
 'partner': 'yes',
 'dependents': 'yes',
 'phoneservice': 'yes',
 'multiplelines': 'yes',
 'internetservice': 'dsl',
 'onlinesecurity': 'yes',
 'onlinebackup': 'no',
 'deviceprotection': 'no',
 'techsupport': 'no',
 'streamingtv': 'yes',
 'streamingmovies': 'yes',
 'contract': 'two_year',
 'paperlessbilling': 'yes',
 'paymentmethod': 'electronic_check',
 'tenure': 67,
 'monthlycharges': 75.5,
 'totalcharges': 5229.45}

In [38]:
train_dict[:1]

[{'gender': 'female',
  'seniorcitizen': 0,
  'partner': 'yes',
  'dependents': 'yes',
  'phoneservice': 'yes',
  'multiplelines': 'yes',
  'internetservice': 'dsl',
  'onlinesecurity': 'yes',
  'onlinebackup': 'no',
  'deviceprotection': 'no',
  'techsupport': 'no',
  'streamingtv': 'yes',
  'streamingmovies': 'yes',
  'contract': 'two_year',
  'paperlessbilling': 'yes',
  'paymentmethod': 'electronic_check',
  'tenure': 67,
  'monthlycharges': 75.5,
  'totalcharges': 5229.45}]

In [39]:
from sklearn.feature_extraction import DictVectorizer

In [40]:
dv = DictVectorizer(sparse= False)

In [41]:
dv.fit(train_dict)

DictVectorizer(sparse=False)

In [42]:
X_train =dv.transform(train_dict)

In [43]:
X_train[0]

array([0.00000e+00, 0.00000e+00, 1.00000e+00, 0.00000e+00, 1.00000e+00,
       1.00000e+00, 0.00000e+00, 0.00000e+00, 1.00000e+00, 0.00000e+00,
       1.00000e+00, 0.00000e+00, 0.00000e+00, 7.55000e+01, 0.00000e+00,
       0.00000e+00, 1.00000e+00, 1.00000e+00, 0.00000e+00, 0.00000e+00,
       0.00000e+00, 0.00000e+00, 1.00000e+00, 0.00000e+00, 1.00000e+00,
       0.00000e+00, 1.00000e+00, 0.00000e+00, 0.00000e+00, 1.00000e+00,
       0.00000e+00, 0.00000e+00, 1.00000e+00, 0.00000e+00, 0.00000e+00,
       0.00000e+00, 1.00000e+00, 0.00000e+00, 0.00000e+00, 1.00000e+00,
       1.00000e+00, 0.00000e+00, 0.00000e+00, 6.70000e+01, 5.22945e+03])

In [44]:
dv.get_feature_names_out()

array(['contract=month-to-month', 'contract=one_year',
       'contract=two_year', 'dependents=no', 'dependents=yes',
       'deviceprotection=no', 'deviceprotection=no_internet_service',
       'deviceprotection=yes', 'gender=female', 'gender=male',
       'internetservice=dsl', 'internetservice=fiber_optic',
       'internetservice=no', 'monthlycharges', 'multiplelines=no',
       'multiplelines=no_phone_service', 'multiplelines=yes',
       'onlinebackup=no', 'onlinebackup=no_internet_service',
       'onlinebackup=yes', 'onlinesecurity=no',
       'onlinesecurity=no_internet_service', 'onlinesecurity=yes',
       'paperlessbilling=no', 'paperlessbilling=yes', 'partner=no',
       'partner=yes', 'paymentmethod=bank_transfer_(automatic)',
       'paymentmethod=credit_card_(automatic)',
       'paymentmethod=electronic_check', 'paymentmethod=mailed_check',
       'phoneservice=no', 'phoneservice=yes', 'seniorcitizen',
       'streamingmovies=no', 'streamingmovies=no_internet_service',

#### Model Kurma

Hedef degiskenimiz Kategorik bir degisken oldugu icin


Cozecegimiz problem SINIFLANDIRMA problemi


Logistic Regresyon Kullanacagiz

In [49]:
from sklearn.linear_model import LogisticRegression

- solver optimizasyon icin kullanilir
- kucuk orneklem icin sklearn liblinear kullanilmasini tavsiye ediyor

In [52]:
model = LogisticRegression(solver='liblinear', random_state=42)

In [ ]:
fit modeli ile modeli egitelim

In [54]:
model.fit(X_train, y_train)

LogisticRegression(random_state=42, solver='liblinear')

#### Model Degerlendirme

- Model iyi mi kotu mu
- Modelin performansini anlamak icin validation  verisini kullanacagiz
- Validation verisine veri onisleme yapilacak

In [55]:
val_dict = df_val[categorical+numerical].to_dict(orient='records')

In [58]:
X_val = dv.transform(val_dict)

In [59]:
y_pred = model.predict_proba(X_val)

In [60]:
y_pred[:5]

array([[0.69754583, 0.30245417],
       [0.96813672, 0.03186328],
       [0.58903798, 0.41096202],
       [0.98736133, 0.01263867],
       [0.25666618, 0.74333382]])

- validation verisinin ilk 5 degeri 0 - 1 araliginda
- churn hedef degiskenin 0 ve 1 olma olasiligi
- 0.5 ten buyukse tahmin 1 0.5 ten kucuk olursa 0 olacak
- ilk musterinin 0 olma olasiligi 83%, 1 olma olasiligi 17%

##### Score metodu ile modelin performasinin olculmesi

In [61]:
model.score(X_val, y_val)

0.808374733853797

- modelin validation versinin uzerindeki performansi 80%

In [62]:
model.score(X_train, y_train)

0.808284023668639

- Modelin egitim versi uzerindeki performansi 80%

##### Iyi bir modelde validation ve egitim performansi uzerindeki modelin dogruluklarinin birbirine ve 1 e yakin olmasi beklenir.

- Modelde ezberleme (overfitting) problemi yok

#### Model Yorumlama

- Kurulan modelde herbir degiskenin bir katsayisi ve bir sabit terimi var
- Iyi bir model en iyi katsayi kombinasyonuna sahip modeldir
- Modelin egitimi sonrasi bu katsayilar Sklearn ile yazdirilabilir

In [66]:
model.intercept_[0] #modelin sabit degeri

-0.1666279293338914

In [68]:
dict(zip(dv.get_feature_names_out(), model.coef_[0].round(3))) # diger degiskenlerin katsayilari

{'contract=month-to-month': 0.726,
 'contract=one_year': -0.074,
 'contract=two_year': -0.819,
 'dependents=no': 0.053,
 'dependents=yes': -0.22,
 'deviceprotection=no': 0.027,
 'deviceprotection=no_internet_service': -0.136,
 'deviceprotection=yes': -0.058,
 'gender=female': -0.084,
 'gender=male': -0.083,
 'internetservice=dsl': -0.377,
 'internetservice=fiber_optic': 0.346,
 'internetservice=no': -0.136,
 'monthlycharges': 0.003,
 'multiplelines=no': -0.272,
 'multiplelines=no_phone_service': 0.127,
 'multiplelines=yes': -0.021,
 'onlinebackup=no': 0.084,
 'onlinebackup=no_internet_service': -0.136,
 'onlinebackup=yes': -0.115,
 'onlinesecurity=no': 0.198,
 'onlinesecurity=no_internet_service': -0.136,
 'onlinesecurity=yes': -0.229,
 'paperlessbilling=no': -0.204,
 'paperlessbilling=yes': 0.037,
 'partner=no': -0.102,
 'partner=yes': -0.065,
 'paymentmethod=bank_transfer_(automatic)': -0.17,
 'paymentmethod=credit_card_(automatic)': -0.138,
 'paymentmethod=electronic_check': 0.224,


- herbir sutunun katsayisi
- pozitif katsayilar hedef degisken ile dogru orantili
- negatif katsayilar hedef degisken ile ters orantili


#### Yeni Veriyi Tahmin Etme

In [70]:
customer ={
 'customerid': '8879-zkjof',
 'gender': 'male',
 'seniorcitizen': 1,
 'partner': 'no',
 'dependents': 'no',
 'phoneservice': 'yes',
 'multiplines': 'no',
 'internetservice': 'dsl',
 'onlinesecurity': 'yes',
 'onlinebackup': 'no',
 'deviceprotection': 'yes',
 'techsupport': 'yes',
 'streamingtv': 'yes',
 'contract': 'one_year',
 'paperlessbilling': 'yes',
 'paymentmethod': 'bank_transfer_(automatic)',
 'monthlycharges': 79.85,
 'totalcharges': 2990.75
}

In [71]:
x_new = dv.transform([customer])

In [72]:
model.predict_proba(x_new)

array([[0.55574526, 0.44425474]])

- Kurdugumuz modelde bu yeni verilerin etiketinin 0 olma ihtimali: 56%
- Kurdugumuz modelde bu yeni verilerin etiketinin 1 olma ihtimali: 44%
- Bu musterinin abonelikten cikma ihtimali abonelikte kalma ihtimaline yakin fakat az

In [73]:
customer2 ={
 'gender': 'female',
 'seniorcitizen': 1,
 'partner': 'no',
 'dependents': 'no',
 'phoneservice': 'yes',
 'multiplines': 'yes',
 'internetservice': 'fiber_optic',
 'onlinesecurity': 'no',
 'onlinebackup': 'no',
 'deviceprotection': 'yes',
 'techsupport': 'yes',
 'streamingtv': 'no',
 'contract': 'month_to_month',
 'paperlessbilling': 'yes',
 'paymentmethod': 'electronic_check',
 'monthlycharges': 85.7,
 'totalcharges': 85.7
}

In [75]:
x_new2 = dv.transform([customer2])

In [76]:
model.predict_proba(x_new2)

array([[0.41708262, 0.58291738]])

- Bu musterinin abonelikten cikma ihtimali 58%

Mulakat(Interview) Sorusu 1:
    
    You would like to create a model to predict a tree's height in the future based on some known parameters and labels from the historical data. What typeof learning algorithm would you choose?

Cozum: 

a. PCA bir boyut indirgeme teknigidir. Denetimsiz ogrenme algoritmasidir. Soruda "labels from the historical data" model kurulmasi isteniyor. Dolayisiyla a yanlis

b. Linear regresyon hedef degiskenin sayisal olma durumunda kullanilir. "a tree's height" sayisal bir degerdir. Olabilir

c. K-Means bir denetimsiz ogrenme algoritmasidir. K-Means icin verde etiketleme olmasina ihtiyac yoktur.

d. Logistic regresyon hedef degiskenin kategorik olmasi durumunda kullanilir. Bu yanlis

Mulakat(Interview) Sorusu 1:
    
    You are building a machine learning model. You have found that the model is performing very well on training data, very poorly on new data. What is the potential isssue in this case?
